### PR TITLE
transmissions as sparse matrix

### PR DESCRIPTION
**Pull request by yongzx**
I have made each `transmissions[t]` to be a sparse matrix of `(i, j, lambda_ij)` and I have updated `get_infection_probas(probas, transmissions)` and `sample_transmissions(self)`.

### CODE REVIEW

File: inference_model.py [hunk 1]
```diff
@@ -8,15 +8,15 @@
...
     """
     - probas[i,s] = P_s^i(t)
-    - transmissions = array/list of i, j, lambda_ij
+    - transmissions = sparse matrix of i, j, lambda_ij
     - infection_probas[i]  = sum_j lambda_ij P_I^j(t)
     """
     N = probas.shape[0]
     infection_probas = np.zeros(N)
     for i in range(N):
+        contact_with_i = transmissions.toarray()[i]
```
Comment (antoinebaker) [inline inference_model.py]: we should replace the for loop by a sparse matrix times vector without passing to a dense array

File: sir_model.py [hunk 2]
```diff
@@ -18,17 +19,14 @@ def get_dummies(states):
...
     """
     infected = (states == 1)
     N = len(states)
     infection_probas = np.zeros(N)
     for i in range(N):
-        rates = np.array([
-            rate for i0, j, rate in transmissions
-            if i0 == i and infected[j]
-        ])
+        rates = transmissions.toarray()[i][infected == True]
```
Comment (antoinebaker) [inline sir_model.py]: same here we should try to replace the whole for loop by a sparse marix times vector

**Review by antoinebaker — CHANGES_REQUESTED** [review]
thx ! but the point of the sparse matrix was to save memory (as does the list) but also to make the computation of get_infection_probas faster. We should replace the inefficient for loops by the appropriate sparse matrix multiply

File: sir_model.py (hunk 2 above)
Comment (yongzx) [reply sir_model.py]: My most recent commit has updated it by 
```python
def get_infection_probas(states, transmissions):
    """
    - states[i] = state of i
    - transmissions = sparse matrix of i, j, lambda_ij
    - infection_probas[i]  = 1 - prod_{j: state==I} [1 - lambda_ij]
    """
    infected = (states == 1)
    N = len(states)
    infected_transmissions = transmissions.multiply(infected)  # element-wise multiplication for masking
    infection_probas = (1 - (1-infected_transmissions.toarray()).prod(axis=1)).squeeze()
    assert len(infection_probas) == N  # sanity check
    return infection_probas
```

File: inference_model.py (hunk 1 above)
Comment (yongzx) [reply inference_model.py]: My most recent commit has updated it by
```python
def get_infection_probas(probas, transmissions):
    """
    - probas[i,s] = P_s^i(t)
    - transmissions = sparse matrix of i, j, lambda_ij
    - infection_probas[i]  = sum_j lambda_ij P_I^j(t)
    """
    N = probas.shape[0]
    infection_probas = transmissions.multiply(probas[:, 1]).toarray().sum(axis=1)
    assert len(infection_probas) == N  # sanity check
    return infection_probas
```

File: sir_model.py [hunk 3]
```diff
@@ -24,10 +24,9 @@ def get_infection_probas(states, transmissions):
     """
     infected = (states == 1)
     N = len(states)
-    infection_probas = np.zeros(N)
-    for i in range(N):
-        rates = transmissions.toarray()[i][infected == True]
-        infection_probas[i] = 1 - np.prod(1 - rates)
+    infected_transmissions = transmissions.multiply(infected)  # element-wise multiplication for masking
+    infection_probas = (1 - (1-infected_transmissions.toarray()).prod(axis=1)).squeeze()
```
Comment (antoinebaker) [inline sir_model.py]: thx a lot ! One last thing, could you use (and test if it works) the trick prod_j (1 - A_ij) = exp(sum_j ln(1-A_ij))  ? that way we can keep a sparse matrix all along, here A_ij = infected_transmissions is sparse and so should be ln(1-A_ij)

Comment (antoinebaker) [reply sir_model.py]: nice, there is even a log1p method for sparse matrices

Comment (yongzx) [reply sir_model.py]: Thanks for the trick! So I have implemented `infection_probas = 1 - np.exp(np.sum(np.log1p(infected_transmissions.multiply(-1)), axis=1))`. It returns the same output as my previous implementation, and I ran it through the Test_inference.ipynb notebook.

**Comment by antoinebaker** [comment]
replaced by 12dca48 with csr matrix vector multiplication, non regression test at fe430a31

**Pull request closed without merging**